File: examples/cpp/licensing_example/licensing_example.cpp
/**
 * Empty example
 */

#include <iostream>
#include <opendaq/opendaq.h>
#include "licensing_example.h"

using namespace daq;

const std::string resourcePath = RESOURCE_PATH;

/*
* A brief example showcasing the use for a licensed module in openDAQ.
* Licensing is currently implemented only in the "LicensingModule", other reference module don't have licensing implemented and behave as usual.
* The module itself requires user authentication (here we use a simple txt file with a password, but in practise an RSA approach
* can be used, where the public key is embedded in the module and the secret key is owned by the relevant user).
* Then the module takes in a license file describing how the module can be used.
* The relevant files are provided with the example. To run the example, the reference quick_start_simulator can be used
* for the input signal.
*/

void tryReadSignalData(const daq::InstancePtr& instance, daq::SignalPtr& inputSignal)
{
    // Sleeping so log output gets flushed before printing..
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    std::cout << "\nAttempting to use a licensed function block...\n" << std::endl;

    // --- Create and connect the unlicensed passthrough function block and read some signals --- //
    const auto fb = instance.addFunctionBlock("LicensingModulePassthrough");

    fb.getInputPorts()[0].connect(inputSignal);

    const uint64_t noOfSamples = 100;

    auto reader = StreamReaderBuilder()
                      .setSignal(fb.getSignals()[0])
                      .setValueReadType(SampleType::Float32)
                      .setDomainReadType(SampleType::Int64)
                      .setReadMode(ReadMode::Scaled)
                      .setReadTimeoutType(ReadTimeoutType::All)
                      .setSkipEvents(true)
                      .build();

    std::this_thread::sleep_for(std::chrono::milliseconds(1000));

    std::vector<float> data(noOfSamples);
    std::vector<int64_t> time(noOfSamples);

    SizeT noOfSamplesRead = noOfSamples;
    auto readerStatus = reader.readWithDomain(data.data(), time.data(), &noOfSamplesRead);
    // Check if we can read the data (otherwise, we might have had a license problem)...
    if (noOfSamplesRead == noOfSamples)
    {
        std::cout << "Read all " << noOfSamplesRead << " samples" << std::endl;
    }
    else
    {
        auto fbStatus = fb.getStatusContainer().getStatusMessage("ComponentStatus");
        if (fbStatus.getLength() == 0)
            fbStatus = "<OK>";

        std::cerr << "Failed to read the expected number of samples. Status of pass-through function block: " << fbStatus
                  << "\nNumber of samples read: " << noOfSamplesRead << std::endl;
    }
}

int main(int /*argc*/, const char* /*argv*/[])
{
    // Create an Instance, loading modules at MODULE_PATH
    const StringPtr path = StringPtr("");
    const ModuleAuthenticatorPtr authenticator = daq::ModuleAuthenticator(path);
    const InstancePtr instance =
        daq::InstanceBuilder().setModuleAuthenticator(authenticator).setLoadAuthenticatedModulesOnly(true).setModulePath(MODULE_PATH).build();

    // Setup your paths here..
    std::string licPath = resourcePath + "license.lic";
    // Make sure you open up the quick_start_simulator example app to get a reference device!
    
    // ------------------- Find and connect to a simulator device ------------------- //
    const auto availableDevices = instance.getAvailableDevices();
    daq::DevicePtr device;
    for (const auto& deviceInfo : availableDevices)
    {
        if (deviceInfo.getName() == "Reference device simulator")
        {
            device = instance.addDevice(deviceInfo.getConnectionString());
            break;
        }
    }

    if (!device.assigned())
    {
        std::cout << "No input device found!" << std::endl;
        return 1;
    }

    std::cout << device.getInfo().getName() << std::endl;

    daq::ChannelPtr inputChannel = device.getChannels()[0];
    daq::SignalPtr inputSignal = inputChannel.getSignals()[0];
    // ------------------------------------------------------------------------------ //

    tryReadSignalData(instance, inputSignal);
    
    // ------------- Find the licensing module ------------------ //
    const auto modules = instance.getModuleManager().getModules();
    auto itFound = std::find_if(modules.begin(),
                                modules.end(),
                                [](const ModulePtr& module) { return module.getModuleInfo().getName() == "LicensingModule"; });

    if (itFound == modules.end())
    {
        std::cerr << "The 'LicensingModule' was not found!" << std::endl << "Press any key to continue...";
        std::cin.get();
        return 1;
    }

    ModulePtr licensingModulePtr = *itFound;
    // ---------------------------------------------------------- //

    // --------------------------- Load the license ------------------------------------------- //
    DictPtr<IString, IString> licenseConfig = licensingModulePtr.getLicenseConfig();
    licenseConfig.set("VendorKey", "my_secret_key");
    licenseConfig.set("LicensePath", licPath);
    Bool succeeded = licensingModulePtr.loadLicense(licenseConfig);

    if (succeeded == false)
    {
        std::cout << "Failed to load license!" << std::endl;
        return 1;
    }
    // ---------------------------------------------------------------------------------------- //

    tryReadSignalData(instance, inputSignal);
    
    std::cout << "Press \"enter\" to exit the application..." << std::endl;
    std::cin.get();
    return 0;
}

File: examples/python/Licensing/licensing_example.py
##
# Function blocks are components that allow can process, generate, or consume signal data. They are bundled in
# openDAQ modules (plugins) where each module can give users access to 0 or more function blocks.
#
# This example prints what function blocks are available within the loaded modules and adds all of them.
##

import opendaq as daq
import sys
sys.path.append("..")

if __name__ == "__main__":
    instance = daq.Instance()

    auth_path = 'authentication_key.txt'
    lic_path = 'license.lic'

    # Create a fresh openDAQ(TM) instance that we will use for all the interactions with the openDAQ(TM) SDK
    instance = daq.Instance()

    # Find and connect to a simulator device
    for device_info in instance.available_devices:
        if device_info.name == 'Reference device simulator':
            device = instance.add_device(device_info.connection_string)
            break
    else:
        # Exit if no device is found
        #exit(0)
        pass

    # Output the name of the added device
    #print(device.info.name)

    licensing_module = None
    for module in instance.module_manager.modules:
        print(module.module_info.id)
        if(module.module_info.id == 'LicensingModule'):
            print("Found!")
            licensing_module = module

    cfg = licensing_module.license_config
    cfg.set_property_value("VendorKey", "my_secret_key")
    cfg.set_property_value("LicensePath", lic_path)

    licensing_module.load_license(cfg)

    # Get the first signal of the first device's channel
    channel = device.channels[0]
    signal = channel.signals[0]

    # Crate an instance of a licensed passthrough function block
    fb = instance.add_function_block("LicensingModulePassthrough")
    fb.input_ports[0].connect(signal)

    # Create an instance of the renderer function block
    renderer = instance.add_function_block('RefFBModuleRenderer')
    renderer.input_ports[0].connect(fb.signals[0])

    input("Press any key to quit...")